File: apps/api/schemas/leaderboards.py
"""
Schemas for the /leaderboards endpoint.

Mirrors the frontend ``types/leaderboards.ts`` shape so a single source
of truth describes the wire format. See ``routers/leaderboards.py`` for
the aggregation logic and ``services/demo_parser_real._build_players``
for where the underlying numbers come from.
"""

from __future__ import annotations

from typing import List, Optional

from pydantic import BaseModel


class LeaderboardEntry(BaseModel):
    """One row of the leaderboard — a player aggregated across demos."""

    rank: int
    steamId: str
    name: str
    clan: Optional[str] = None
    demos: int
    rounds: int
    # Rating + per-round rates, all computed from the raw counts so
    # the values stay correct across multiple demos with different
    # round counts.
    rating: float
    adr: float
    kd: float
    kpr: float
    dpr: float
    kast: float
    udr: float
    far: float
    # Raw counters (useful for tooltips / debugging).
    kills: int
    deaths: int
    assists: int
    headshots: int


class LeaderboardFilters(BaseModel):
    """Echo back which filters were applied — handy for the UI."""

    map: Optional[str] = None
    since: Optional[str] = None
    minRounds: int
    limit: int


class LeaderboardResponse(BaseModel):
    filters: LeaderboardFilters
    entries: List[LeaderboardEntry]
    total: int

File: apps/web/types/leaderboards.ts
// Wire shape for /leaderboards — mirrors apps/api/schemas/leaderboards.py.
// Keep the two in sync (TS field names match Pydantic camelCase aliases).

export interface LeaderboardEntry {
  rank: number;
  steamId: string;
  name: string;
  clan: string | null;
  demos: number;
  rounds: number;
  // Rating + per-round rates. All computed server-side from raw
  // counters so the same demo aggregated across map filters returns
  // self-consistent numbers.
  rating: number;
  adr: number;
  kd: number;
  kpr: number;
  dpr: number;
  kast: number;
  udr: number;
  far: number;
  // Raw counters (for tooltips / drill-downs).
  kills: number;
  deaths: number;
  assists: number;
  headshots: number;
}

export interface LeaderboardFilters {
  map: string | null;
  since: string | null;
  minRounds: number;
  limit: number;
}

export interface LeaderboardResponse {
  filters: LeaderboardFilters;
  entries: LeaderboardEntry[];
  total: number;
}

export interface LeaderboardQuery {
  map?: string;
  since?: string;
  minRounds?: number;
  limit?: number;
}
